File: node/scripts/msgpack_echo_client.js
const { Peer } = require('..')
const { encode, decode } = require('@msgpack/msgpack')

const url = process.argv[2] || 'tcp://127.0.0.1:24010'
const method = process.argv[3] || 'MsgPack/Echo'
const msg = process.argv[4] || 'Hello'

;(async () => {
  const p = await Peer.connect(url)
  const resp = await p.callRaw(method, Buffer.from(encode({ message: msg })))
  const obj = decode(resp)
  process.stdout.write(`reply: ${obj.message}\n`)
})().catch((e) => {
  console.error(e)
  process.exit(1)
})

File: rust/examples/msgpack_echo_client.rs
use std::env;

use bytes::Bytes;
use forpc::RpcPeer;

#[derive(serde::Serialize, serde::Deserialize)]
struct EchoMsg {
    message: String,
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    let mut args = env::args().skip(1);
    let url = args
        .next()
        .unwrap_or_else(|| "tcp://127.0.0.1:24010".to_string());
    let method = args.next().unwrap_or_else(|| "MsgPack/Echo".to_string());
    let msg = args.next().unwrap_or_else(|| "Hello".to_string());

    let peer = RpcPeer::connect_with_retry(&url, 50).await?;

    let peer_clone = peer.clone();
    tokio::spawn(async move {
        let _ = peer_clone.serve().await;
    });

    let req = EchoMsg { message: msg };
    let payload = rmp_serde::to_vec(&req)?;

    let resp = peer.call_raw(&method, Bytes::from(payload)).await?;

    let reply: EchoMsg = rmp_serde::from_slice(&resp)?;
    println!("reply: {}", reply.message);
    Ok(())
}
